### PR TITLE
operator/ingress: introduce tlsSecurityProfile field

### PR DESCRIPTION
**Pull request by ironcladlou**
Support tlsSecurityProfile configuration on ingresscontroller.

Part of the [TLS Config for Externally Facing Services](https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/tls-config.md) enhancement.

@openshift/sig-network-edge @deads2k @enj

### CODE REVIEW

File: operator/v1/types_ingress.go [hunk 1]
```diff
@@ -332,6 +340,10 @@ type IngressControllerStatus struct {
 	//     * DNS records have been successfully created.
 	//   - False if any of those conditions are unsatisfied.
 	Conditions []OperatorCondition `json:"conditions,omitempty"`
+
+	// tlsSecurityProfile is the observed TLS connection configuration.
+	// +optional
+	TLSSecurityProfile *configv1.TLSSecurityProfile `json:"tlsSecurityProfile,omitempty"`
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: Shouldn't this be `*configv1.TLSProfileSpec`?

File: operator/v1/types_ingress.go [hunk 2]
```diff
@@ -125,6 +127,12 @@ type IngressControllerSpec struct {
 	//
 	// +optional
 	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
+
+	// tlsSecurityProfile specifies settings for TLS connections for externally exposed servers.
+	//
+	// If unset, a default (which may change between releases) is chosen.
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: This statement is ambiguous.  The default value for a new `IngressController` may change between releases, but does that change the effective value for an `IngressController` that was created with the default value?

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Hm. I think so.

File: operator/v1/types_ingress.go [hunk 3]
```diff
@@ -332,6 +340,10 @@ type IngressControllerStatus struct {
 	//     * DNS records have been successfully created.
 	//   - False if any of those conditions are unsatisfied.
 	Conditions []OperatorCondition `json:"conditions,omitempty"`
+
+	// tlsSecurityProfile is the observed TLS connection configuration.
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: Could we change "observed" to "effective" (or rephrase as "the TLS connection configuration that is in effect")?

File: operator/v1/types_ingress.go [hunk 4]
```diff
@@ -125,6 +127,12 @@ type IngressControllerSpec struct {
 	//
 	// +optional
 	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
+
+	// tlsSecurityProfile specifies settings for TLS connections for externally exposed servers.
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: "servers"→"IngressControllers", or for better consistency in terminology, how about "externally published IngressControllers"?

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Doesn't this mean we also need to add `TLSSecurityProfile` to `ingresses.config.openshift.io` like https://github.com/openshift/api/blob/master/config/v1/types_apiserver.go#L49 so that we can say our default is based on the cluster singleton?

File: operator/v1/types_ingress.go (hunk 2 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: See https://github.com/openshift/api/pull/454#discussion_r330267806

File: operator/v1/types_ingress.go [hunk 5]
```diff
@@ -332,6 +340,10 @@ type IngressControllerStatus struct {
 	//     * DNS records have been successfully created.
 	//   - False if any of those conditions are unsatisfied.
 	Conditions []OperatorCondition `json:"conditions,omitempty"`
+
+	// tlsProfile is the TLS connection configuration that is in effect.
+	// +optional
+	TLSProfile *configv1.TLSProfileSpec `json:"tlsProfile,omitempty"`
```
Comment (danehans) [inline operator/v1/types_ingress.go]: What about reflecting the TLS profile details in status? Something like:
```
TLSProfile *configv1.TLSProfileSpec `json:"tlsProfileSpec,omitempty"`
```

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (Miciah) [reply operator/v1/types_ingress.go]: I don't see how that follows.  However, it would make sense to have some cluster-wide default TLS configuration.  The same default could be used for ingress, the API, and any other component where it would make sense.

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Added config to ingress.config.openshift.io.

File: operator/v1/types_ingress.go (hunk 2 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: I now reference the ingress cluster config as the source of defaulting.

File: operator/v1/types_ingress.go (hunk 4 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 5 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Done

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: It follows the design laid out in https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/tls-config.md and applied to APIServerSpec. Where else could the defaults possibly come from?

Comment (Miciah) [reply operator/v1/types_ingress.go]: > It follows the design laid out in https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/tls-config.md and applied to APIServerSpec.

I agree that a requirement to have a cluster-wide setting follows from https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/tls-config.md (specifically, the statement, "This should be a shared, top level configuration amongst the affected components with a local override for ingress controllers").  I don't see how it follows from using `*configv1.TLSProfileSpec` for the value of this field.  However, in this case, it doesn't matter if we use different paths to arrive at the same conclusion.

> Where else could the defaults possibly come from?

Hard-coded defaults.  But using a cluster singleton is more obvious, flexible, and maintainable.

File: operator/v1/types_ingress.go (hunk 2 above)
Comment (Miciah) [reply operator/v1/types_ingress.go]: That does not change anything.  If I create an ingresscontroller with the default TLS profile and then update the TLS profile in the ingress cluster config, does the ingresscontroller continue to use the old default, or should the ingress operator update the ingresscontroller's effective TLS profile to the new default value?

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Good question. I think that question applies to more than just this field.

I guess the semantics are up to us to define and document here on a case by case basis, because there is no coherent or enforceable rule applied broadly which would just answer the question for us.

How about:

>If unset, the default is based on the ingresses.config.openshift.io/cluster resource.
>Note that if the default is used, and the ingresses.config.openshift.io/cluster resource
>diverges from the ingresscontroller's effective default, the ingresscontroller will continue
>to use the defaults applied at the time of creation.

Comment (Miciah) [reply operator/v1/types_ingress.go]: Yeah, I like that because (1) it is consistent with other fields in `IngressControllerSpec`, (2) it avoids possibly confusing action at a distance, and (3) it is easiest to implement.

File: operator/v1/types_ingress.go [hunk 6]
```diff
@@ -125,6 +127,13 @@ type IngressControllerSpec struct {
 	//
 	// +optional
 	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
+
+	// tlsSecurityProfile specifies settings for TLS connections for ingresscontrollers.
+	//
+	// If unset, the default is based on the ingresses.config.openshift.io/cluster resource.
+	//
+	// +optional
+	TLSSecurityProfile *configv1.TLSSecurityProfile `json:"tlsSecurityProfile,omitempty"`
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: Is `spec.tlsSecurityProfile` mutable?

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Thinking about this some more, another option would be reading effective values from the [static mappings](https://github.com/openshift/api/blob/master/config/v1/types_tlssecurityprofile.go#L274) and never persisting them. In that case status would in fact contain a `TLSSecurityProfile` to avoid persisting the statically mapped spec data. I would like to see how @deads2k is planning on doing this for the APIServer.

File: operator/v1/types_ingress.go (hunk 6 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Maybe given we go out of our way to call out things which are immutable, we can assume that it's mutable? I think generally users expect to be able to change spec fields and have those changes reflected.

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: To further elaborate on what I'm saying:

Input:

```yaml
apiVersion: config.openshift.io/v1
kind: Ingress
spec:
  tlsSecurityProfile:
    type: Intermediate

apiVersion: operator.openshift.io/v1
kind: IngressController
spec:
  tlsSecurityProfile: nil
```

Output:

```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
spec:
  tlsSecurityProfile: nil
status:
  tlsProfileSpec:
    # from https://github.com/openshift/api/blob/master/config/v1/types_tlssecurityprofile.go#L313
    ciphers: # ...
    tlsVersion: # ...
    dhParamSize: # ...
```

Now, just looking at the ingresscontroller, what profile is it using? Assume the cluster config changed, so the original cluster default that led to the status is gone.

Instead:

```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
spec:
  tlsSecurityProfile: nil
status:
  tlsSecurityProfile:
    type: Intermediate
```

The ingress-operator uses the static mappings in `types_tlssecurityprofile.go` to convert this into a TLSSecuritySpec at runtime.

I suspect there are reasons for and against both ways I'm not considering so I wanted to clarify for discussion.

Comment (Miciah) [reply operator/v1/types_ingress.go]: I dislike the latter option because (1) the TLS profile of an existing ingresscontroller is changed _implicitly_ when the "Intermediate" profile changes, (2) the status is a lie during the period of time between when the definition of "Intermediate" changes and when the ingress operator reconciles the ingresscontroller, and (3) the status no longer says _explicitly_ what ciphers and TLS versions are enabled.

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: >I dislike the latter option because (1) the TLS profile of an existing ingresscontroller is changed implicitly when the "Intermediate" profile changes, (2) the status is a lie during the period of time between when the definition of "Intermediate" changes and when the ingress operator reconciles the ingresscontroller, and (3) the status no longer says explicitly what ciphers and TLS versions are enabled.

Do you agree that the former way causes information loss? (e.g. the user's explicit intent to use the "Intermediate" profile.) If so, does it matter?

If the ingresscontroller is changed from Intermediate to Modern and the client misses the update and clears its cache, it won't have any knowledge of the Intermediate state (as it can't be inferred from status), but we should be level-based anyway.

I'm trying to think of a useful way to document the semantics... like, the profile type you specify is used to stamp down a generic point-in-time representation of the specific profile, and if you change the profile type on the existing ingresscontroller, it's possible a different config is stamped down even if the type is the same (i.e. we updated what "Intermediate" is).

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: >I'm trying to think of a useful way to document the semantics... like, the profile type you specify is used to stamp down a generic point-in-time representation of the specific profile, and if you change the profile type on the existing ingresscontroller, it's possible a different config is stamped down even if the type is the same (i.e. we updated what "Intermediate" is).

My example here was wrong.. imagine if you specify Intermediate and then upgrade to a release where we change what Intermediate consists of. Do we rebuild your status and roll out automatically? If not, when, and how would we prevent that from happening the next time we reconcile for any other reason.

(Related to the mutability question?)

Comment (Miciah) [reply operator/v1/types_ingress.go]: > Do you agree that the former way causes information loss? (e.g. the user's explicit intent to use the "Intermediate" profile.) If so, does it matter?

The intent is in `spec`, so what is lost?

> If the ingresscontroller is changed from Intermediate to Modern and the client misses the update and clears its cache, it won't have any knowledge of the Intermediate state (as it can't be inferred from status), but we should be level-based anyway.

Which client are you talking about?  The operator's client, the client with which the administrator set `spec.tlsSecurityProfile`, or something else?

> My example here was wrong.. imagine if you specify Intermediate and then upgrade to a release where we change what Intermediate consists of. Do we rebuild your status and roll out automatically? If not, when, and how would we prevent that from happening the next time we reconcile for any other reason.

If `spec.tlsSecurityProfile` is mutable, I think the operator will have to updating the effective profile in this scenario.  Anything else would be an awful mess.  So let's tell the user to specify an explicit custom profile if ey needs it and otherwise accept that the profile ey specifies is subject to change.

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: >The intent is in spec, so what is lost?

The prior intent.

>Which client are you talking about? The operator's client, the client with which the administrator set spec.tlsSecurityProfile, or something else?

Operator. But as I said (and taken with what you say next), I don't think it matters because we're talking about a level driven approach — the old intent doesn't matter. So never mind all that.

>If spec.tlsSecurityProfile is mutable, I think the operator will have to update the effective profile in this scenario. Anything else would be an awful mess. So let's tell the user to specify an explicit custom profile if ey needs it and otherwise accept that the profile ey specifies is subject to change.

👍 Okay, I think we're on the same page.

File: operator/v1/types_ingress.go (hunk 2 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Updated docs

File: operator/v1/types_ingress.go [hunk 7]
```diff
@@ -131,6 +131,15 @@ type IngressControllerSpec struct {
 	// tlsSecurityProfile specifies settings for TLS connections for ingresscontrollers.
 	//
 	// If unset, the default is based on the ingresses.config.openshift.io/cluster resource.
+	// If the default is used, and the ingresses.config.openshift.io/cluster resource
+	// diverges from the ingresscontroller's effective default, the ingresscontroller will
+	// continue to use the defaults applied at the time of creation.
+	//
+	// Note that when using the Old, Intermediate, and Custom profile types, the effective
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: "Custom"→"Modern".

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: D'oh, fixed

File: operator/v1/types_ingress.go [hunk 8]
```diff
@@ -131,6 +131,15 @@ type IngressControllerSpec struct {
 	// tlsSecurityProfile specifies settings for TLS connections for ingresscontrollers.
 	//
 	// If unset, the default is based on the ingresses.config.openshift.io/cluster resource.
+	// If the default is used, and the ingresses.config.openshift.io/cluster resource
+	// diverges from the ingresscontroller's effective default, the ingresscontroller will
+	// continue to use the defaults applied at the time of creation.
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: I think based on the most recent discussion, this is false: the operator will eschew whatever has been persisted in the ingresscontroller's status and reconcile the effective TLS profile.

Comment (ironcladlou) [reply operator/v1/types_ingress.go]: I agree. Removed this entire section.

File: config/v1/types_ingress.go [hunk 9]
```diff
@@ -32,6 +32,12 @@ type IngressSpec struct {
 	//
 	// Once set, changing domain is not currently supported.
 	Domain string `json:"domain"`
+
+	// tlsSecurityProfile specifies settings for TLS connections for ingresscontrollers.
+	//
+	// If unset, a default (which may change between releases) is chosen.
+	// +optional
+	TLSSecurityProfile *TLSSecurityProfile `json:"tlsSecurityProfile,omitempty"`
```
Comment (deads2k) [inline config/v1/types_ingress.go]: I don't think we should do this.  Having a default for ingresscontrollers match the default for the APIServer makes a cluster-wide default work.  Then you only need the value on your operator in this pull.

File: operator/v1/types_ingress.go [hunk 10]
```diff
@@ -125,6 +127,19 @@ type IngressControllerSpec struct {
 	//
 	// +optional
 	NodePlacement *NodePlacement `json:"nodePlacement,omitempty"`
+
+	// tlsSecurityProfile specifies settings for TLS connections for ingresscontrollers.
+	//
+	// If unset, the default is based on the ingresses.config.openshift.io/cluster resource.
```
Comment (deads2k) [inline operator/v1/types_ingress.go]: based on `oc get apiservers.config.openshift.io/cluster` resource

File: operator/v1/types_ingress.go [hunk 11]
```diff
@@ -332,6 +347,10 @@ type IngressControllerStatus struct {
 	//     * DNS records have been successfully created.
 	//   - False if any of those conditions are unsatisfied.
 	Conditions []OperatorCondition `json:"conditions,omitempty"`
+
+	// tlsProfile is the TLS connection configuration that is in effect.
```
Comment (deads2k) [inline operator/v1/types_ingress.go]: When you have multiple pods at different TLS profiles, what value will this show me?

File: config/v1/types_ingress.go (hunk 9 above)
Comment (ironcladlou) [reply config/v1/types_ingress.go]: Are you saying apiserver.config.openshift.io is the cluster default ingress should honor in this case?

File: operator/v1/types_ingress.go (hunk 11 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: I'm not sure what you mean here, can you rephrase, maybe with an example in terms of ingresscontrollers?

File: config/v1/types_ingress.go (hunk 9 above)
Comment (ironcladlou) [reply config/v1/types_ingress.go]: @Miciah what's your take here?

Comment (ironcladlou) [reply config/v1/types_ingress.go]: I have no objection, just surprised we had this misunderstanding to begin with and makes me wonder what else I'm missing.

Comment (Miciah) [reply config/v1/types_ingress.go]: I find it surprising that `apiserver.config.openshift.io/cluster` has the default value for ingresscontrollers (and whatever future components need configuration for TLS profile).

File: operator/v1/types_ingress.go (hunk 11 above)
Comment (deads2k) [reply operator/v1/types_ingress.go]: ingresscontroller leads to 3 pods total.  user changes cipher suites and you being rolling out.  As you replace one pod at a time, the ingresscontroller has 2 old pods and 1 new pod, then gets stuck. What is the value of this field?

File: config/v1/types_ingress.go (hunk 9 above)
Comment (deads2k) [reply config/v1/types_ingress.go]: It is not reasonable to have two cluster-wide default values for this. We should have one.  Ingresses may or may not be used (a user could set up a different mechanism), but APIServers will always be present, so we placed it on apiservers.config.openshift.io.  Making a user specify it twice in the common case isn't friendly or useful to them.

File: operator/v1/types_ingress.go (hunk 11 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Well, it would be either the old or the new cipher config depending on how we implement the reconciliation. It sounds like you already have an idea of which it should be and why — if so can you elaborate?

If the value is the new config, and we haven't fully converged on that value, is status misleading?

If the value is the old config, and the spec value is nil (indicating the use of runtime defaults), status doesn't reflect the state of the new pods until full convergence — any problems there?

File: config/v1/types_ingress.go (hunk 9 above)
Comment (ironcladlou) [reply config/v1/types_ingress.go]: Sounds fine to me — I had originally interpreted the design as including a cluster default to drive multiple components, but the API being scoped to the apiserver resource is what surprised me. As long as the docs are clear about how the defaults propagate I don't see an issue. @Miciah, WDYT?

Comment (Miciah) [reply config/v1/types_ingress.go]: > Ingresses may or may not be used (a user could set up a different mechanism)

We're talking about `ingress.config/cluster`, right?  That should always be present, and it should have ingress-related cluster configuration.

Comment (ironcladlou) [reply config/v1/types_ingress.go]: @deads2k do you have more to say here? It sounds like @Miciah advocates for what I have already, and you're advocating for something else, and I need to know how we come to consensus. Is there anybody else you know of with an opinion?

Comment (ironcladlou) [reply config/v1/types_ingress.go]: @deads2k @knobunc @Miciah I have removed the `ingress.config.openshift.io` change. The operator can use `apiservers.config.openshift.io` instead, and we can discuss that topic separately.

File: operator/v1/types_ingress.go (hunk 10 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: Fixed

File: operator/v1/types_ingress.go (hunk 11 above)
Comment (ironcladlou) [reply operator/v1/types_ingress.go]: @deads2k I need a reply to this. Are you suggesting a specific change? If not, I need to know so we can move on.

**Comment by Miciah** [comment]
I still prefer `ingress.config/cluster` for default cluster ingress TLS config, but we need something.
/lgtm

File: operator/v1/types_ingress.go (hunk 11 above)
Comment (wking) [reply operator/v1/types_ingress.go]: > Well, it would be either the old or the new cipher config depending on how we implement the reconciliation.

Could be a third value too:

1. `spec` and `status` at A, all pods at A, life is good.
2. Bump `spec` to B, some pods upgrade to B.
3. Bump `spec` to C, some pods upgrade to C.  Still have some pods at both A and B.

But if we bump the `status` value when we complete a rollout, then folks can use "`status` diverges from `spec`" to see that that particular value is still progressing, and Progressing status could be exposed in a human-readable way in a Progressing condition.  So I don't think this blocks this PR unless we want to bump the docs to change "is in effect" -> "was most recently reconciled" or some such.
